src/bin/common/solving_loop.ml

```diff
@@ -682,61 +767,79 @@ let main () =
+            Expr.print expr
+        end
+      else
+        Api.FE.optimize ~loc (expr, is_max) Api.env
+    in st
   in
 
   let handle_get_objectives (_args : DStd.Expr.Term.t list) st =
-    let () =
-      if Options.get_interpretation () then
-        match State.get partial_model_key st with
-        | Some Model ((module SAT), partial_model) ->
-          let objectives = SAT.get_objectives partial_model in
-          begin
-            match objectives with
-            | Some o ->
-              if not @@ Objective.Model.has_no_limit o then
-                warning "Some objectives cannot be fulfilled";
-              Objective.Model.pp (Options.Output.get_fmt_regular ()) o
-            | None ->
-              recoverable_error "No objective generated"
-          end
-        | None ->
-          recoverable_error
-            "Model generation is disabled (try --produce-models)"
-    in
-    st
+    if Options.get_interpretation () then
+      match State.get partial_model_key st with
+      | Some Model ((module SAT), partial_model) ->
+        let objectives = SAT.get_objectives partial_model in
+        begin
+          match objectives with
+          | Some o ->
+            if not @@ Objective.Model.has_no_limit o then
+              warning "Some objectives cannot be fulfilled";
+            Objective.Model.pp (Options.Output.get_fmt_regular ()) o
+          | None ->
+            recoverable_error "No objective generated"
+        end
+      | None ->
+        recoverable_error
+          "Model generation is disabled (try --produce-models)"
   in
 
   let handle_custom_statement loc id args st =
     let args = List.map Dolmen_type.Core.Smtlib2.sexpr_as_term args in
     let logic_file = State.get State.logic_file st in
     let st, terms = Typer.terms st ~input:(`Logic logic_file) ~loc args in
+    let loc =
+      let dloc_file = (State.get State.logic_file st).loc in
+      dl_to_ael dloc_file loc
+    in
     match id, terms.ret with
-    | Dolmen.Std.Id.{name = Simple "minimize"; _}, [term] ->
+    | Dolmen.Std.Id.{name = Simple ("minimize" as name_base); _}, [term] ->
       cmd_on_modes st [Assert] "minimize";
-      handle_optimize_stmt ~is_max:false loc id term st
-    | Dolmen.Std.Id.{name = Simple "maximize"; _}, [term] ->
+      let expr =
+        D_cnf.make_expr
+          ~loc
+          ~name_base
+          ~toplevel:true
+          ~decl_kind:Expr.Dobjective
+          term
+      in
+      handle_optimize_stmt ~is_max:false loc expr st
+    | Dolmen.Std.Id.{name = Simple ("maximize" as name_base); _}, [term] ->
       cmd_on_modes st [Assert] "maximize";
-      handle_optimize_stmt ~is_max:true loc id term st
+      let expr =
+        D_cnf.make_expr
+          ~loc
+          ~name_base
+          ~toplevel:true
+          ~decl_kind:Expr.Dobjective
+          term
+      in
+      handle_optimize_stmt ~is_max:true loc expr st
```

Comment:
I guess these two cases can be factorized with `("minimize" | "maximize") as name_base` and `let is_max = match name_base with "minimize" -> false | "maximize" -> true | _ -> assert false in`.